src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -27,56 +27,166 @@ internal class SemanticTokensHelpers
+                public const string ExcludedCode = "excluded code";
+                public const string WhiteSpace = "whitespace";
+                public const string Text = "text";
+
+                internal const string ReassignedVariable = "reassigned variable";
+                public const string StaticSymbol = "static symbol";
+
+                public const string PreprocessorText = "preprocessor text";
+                public const string Punctuation = "punctuation";
+                public const string StringEscapeCharacter = "string - escape character";
+#endif
+
+            }).ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+
+        public static Dictionary<string, string> GetTokenTypeMap(ClientCapabilities? capabilities)
+            => capabilities != null && capabilities.HasVisualStudioLspCapability()
+                ? s_VSClassificationTypeToSemanticTokenTypeMap
+                : s_pureLspClassificationTypeToSemanticTokenTypeMap;
+
+        public static ImmutableArray<string> GetCustomTokenTypes(ClientCapabilities? capabilities)
+            => GetCustomTokenTypes(GetTokenTypeMap(capabilities));
+
+        private static ImmutableArray<string> GetCustomTokenTypes(Dictionary<string, string> tokenMap)
+            => ClassificationTypeNames.AllTypeNames
+                .Where(type => !tokenMap.ContainsKey(type) && !ClassificationTypeNames.AdditiveTypeNames.Contains(type))
+                .Order()
+                .ToImmutableArray();
+
+        /// <summary>
+        /// Gets all the supported token types for the provided <paramref name="capabilities"/>.  If <paramref
+        /// name="capabilities"/> this will be the core set of LSP token types that roslyn supports.  Depening on the
+        /// capabilities passed in this may be a different set (for example, VS supports more semantic token types).
+        /// </summary>
+        /// <param name="capabilities"></param>
+        /// <returns></returns>
+        public static ImmutableArray<string> GetAllTokenTypes(ClientCapabilities? capabilities)
+            => SemanticTokenTypes.AllTypes.Concat(GetCustomTokenTypes(capabilities)).ToImmutableArray();
+
+        public static ImmutableArray<string> LegacyGetAllTokenTypesForRazor()
+            => SemanticTokenTypes.AllTypes.Concat(GetCustomTokenTypes(s_VSClassificationTypeToSemanticTokenTypeMap)).ToImmutableArray();
+
+        public static Dictionary<string, int> GetTokenTypeToIndex(ClientCapabilities capabilities)
         {
-            // Computes the mapping between a LSP token type and its respective index recognized by LSP.
-            TokenTypeToIndex = new Dictionary<string, int>();
-            var index = 0;
-            foreach (var lspTokenType in LSP.SemanticTokenTypes.AllTypes)
+            if (s_tokenTypeToIndex == null)
             {
-                TokenTypeToIndex.Add(lspTokenType, index);
-                index++;
-            }
+                s_tokenTypeToIndex = new Dictionary<string, int>();
```

Comment:
>s_tokenTypeToIndex = new Dictionary<string, int>(); [](http://example.com/codeflow?start=16&length=51)

No need to worry about simultaneous calls to GetTokenTypeToIndex, right?

Reply:
Hrmm... Actually, yes there is.  these requests can run concurrently on the server if we're asked about multiple files.  I'll fix.

Reply:
fixed.